File: src/main/java/org/example/kalkulationsprogramm/repository/ProjektDokumentRepository.java
package org.example.kalkulationsprogramm.repository;

import java.util.List;
import java.util.Optional;

import org.example.kalkulationsprogramm.domain.ProjektDokument;
import org.example.kalkulationsprogramm.domain.ProjektGeschaeftsdokument;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

@Repository
public interface ProjektDokumentRepository extends JpaRepository<ProjektDokument, Long> {
  List<ProjektDokument> findByProjektId(Long projektId);

  @Query("SELECT g FROM ProjektGeschaeftsdokument g")
  List<ProjektGeschaeftsdokument> findAllGeschaeftsdokumente();

  @Query("""
      SELECT g FROM ProjektGeschaeftsdokument g
      WHERE g.bezahlt = false
        AND (
              LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%mahn%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%erinnerung%'
        )
      """)
  List<ProjektGeschaeftsdokument> findOffeneGeschaeftsdokumente();

  @Query("SELECT g FROM ProjektGeschaeftsdokument g WHERE LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%' AND g.rechnungsdatum BETWEEN :start AND :end")
  List<ProjektGeschaeftsdokument> findGeschaeftsdokumenteByRechnungsdatumBetween(java.time.LocalDate start,
      java.time.LocalDate end);

  @Query("SELECT g FROM ProjektGeschaeftsdokument g WHERE g.projekt.id = :projektId AND g.geschaeftsdokumentart = 'Rechnung'")
  List<ProjektGeschaeftsdokument> findRechnungenByProjektId(@Param("projektId") Long projektId);

  @Query("select g.projekt.id, g.dokumentid from ProjektGeschaeftsdokument g where g.projekt.id in :ids")
  List<Object[]> findDokumentIdsByProjektIds(List<Long> ids);

  /**
   * Liefert für eine Liste von Projekt-IDs ein Mapping
   * [ProjektGeschaeftsdokument.id, Projekt.id], damit der DokumentFreigabeService
   * den jüngsten Freigabe-Status pro Projekt finden kann.
   */
  @Query("SELECT g.id, g.projekt.id FROM ProjektGeschaeftsdokument g WHERE g.projekt.id IN :projektIds")
  List<Object[]> findGeschaeftsdokumentIdMappingByProjektIds(@Param("projektIds") List<Long> projektIds);

  /**
   * Prüft ob ein Geschäftsdokument mit der Dokumentnummer bereits existiert.
   */
  @Query("SELECT CASE WHEN COUNT(g) > 0 THEN true ELSE false END FROM ProjektGeschaeftsdokument g WHERE g.dokumentid = :dokumentid")
  boolean existsByDokumentid(@Param("dokumentid") String dokumentid);

  /**
   * Prüft ob es noch unbezahlte Rechnungen für ein Projekt gibt.
   */
  @Query("""
      SELECT CASE WHEN COUNT(g) > 0 THEN true ELSE false END
      FROM ProjektGeschaeftsdokument g
      WHERE g.projekt.id = :projektId
        AND g.bezahlt = false
        AND (
              LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%mahn%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%erinnerung%'
        )
      """)
  boolean existsOffenePostenByProjektId(@Param("projektId") Long projektId);

  Optional<ProjektDokument> findByGespeicherterDateiname(String gespeicherterDateiname);

  Optional<ProjektDokument> findByGespeicherterDateinameIgnoreCase(String gespeicherterDateiname);

  Optional<ProjektDokument> findByOriginalDateinameIgnoreCase(String originalDateiname);
}

File: src/main/kotlin/org/example/kalkulationsprogramm/repository/ProjektDokumentRepository.kt
package org.example.kalkulationsprogramm.repository

import java.time.LocalDate
import java.util.Optional
import org.example.kalkulationsprogramm.domain.ProjektDokument
import org.example.kalkulationsprogramm.domain.ProjektGeschaeftsdokument
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface ProjektDokumentRepository : JpaRepository<ProjektDokument, Long> {
    fun findByProjektId(projektId: Long?): List<ProjektDokument>

    @Query("SELECT g FROM ProjektGeschaeftsdokument g")
    fun findAllGeschaeftsdokumente(): List<ProjektGeschaeftsdokument>

    @Query(
        """
      SELECT g FROM ProjektGeschaeftsdokument g
      WHERE g.bezahlt = false
        AND (
              LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%mahn%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%erinnerung%'
        )
      """,
    )
    fun findOffeneGeschaeftsdokumente(): List<ProjektGeschaeftsdokument>

    @Query("SELECT g FROM ProjektGeschaeftsdokument g WHERE LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%' AND g.rechnungsdatum BETWEEN :start AND :end")
    fun findGeschaeftsdokumenteByRechnungsdatumBetween(
        start: LocalDate,
        end: LocalDate,
    ): List<ProjektGeschaeftsdokument>

    @Query("SELECT g FROM ProjektGeschaeftsdokument g WHERE g.projekt.id = :projektId AND g.geschaeftsdokumentart = 'Rechnung'")
    fun findRechnungenByProjektId(@Param("projektId") projektId: Long?): List<ProjektGeschaeftsdokument>

    @Query(
        """
      SELECT g FROM ProjektGeschaeftsdokument g
      WHERE g.projekt.id = :projektId
        AND g.mahnstufe IS NOT NULL
      """,
    )
    fun findMahnungenByProjektId(@Param("projektId") projektId: Long?): List<ProjektGeschaeftsdokument>

    @Query("select g.projekt.id, g.dokumentid from ProjektGeschaeftsdokument g where g.projekt.id in :ids")
    fun findDokumentIdsByProjektIds(ids: List<Long>): List<Array<Any>>

    @Query("SELECT g.id, g.projekt.id FROM ProjektGeschaeftsdokument g WHERE g.projekt.id IN :projektIds")
    fun findGeschaeftsdokumentIdMappingByProjektIds(
        @Param("projektIds") projektIds: List<Long>,
    ): List<Array<Any>>

    @Query("SELECT CASE WHEN COUNT(g) > 0 THEN true ELSE false END FROM ProjektGeschaeftsdokument g WHERE g.dokumentid = :dokumentid")
    fun existsByDokumentid(@Param("dokumentid") dokumentid: String): Boolean

    @Query(
        """
      SELECT CASE WHEN COUNT(g) > 0 THEN true ELSE false END
      FROM ProjektGeschaeftsdokument g
      WHERE g.projekt.id = :projektId
        AND g.bezahlt = false
        AND (
              LOWER(g.geschaeftsdokumentart) LIKE '%rechnung%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%mahn%'
           OR LOWER(g.geschaeftsdokumentart) LIKE '%erinnerung%'
        )
      """,
    )
    fun existsOffenePostenByProjektId(@Param("projektId") projektId: Long?): Boolean

    fun findByGespeicherterDateiname(gespeicherterDateiname: String): Optional<ProjektDokument>

    fun findByGespeicherterDateinameIgnoreCase(gespeicherterDateiname: String): Optional<ProjektDokument>

    fun findByOriginalDateinameIgnoreCase(originalDateiname: String): Optional<ProjektDokument>
}
